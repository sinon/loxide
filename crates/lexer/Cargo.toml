[package]
name = "lexer"
version = "0.1.0"
edition = "2024"

[dependencies]
miette = { workspace = true }


[dev-dependencies]
rstest = "0.23.0"
insta = { version = "1.41.1", features = ["yaml"] }
