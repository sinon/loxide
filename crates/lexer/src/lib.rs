//! Lexer
//!
//! Responsible for transforming a given input str into a Iterator of `Result<Token>`
#![allow(clippy::too_many_lines)]

use miette::{Error, LabeledSpan, Result, miette};

#[derive(Clone, Debug)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

/// `Token` is formed of a token type (`TokenType`) and a reference to a str in the input string
#[derive(Clone, Debug)]
pub struct Token {
    /// The `TokenType` of `Token`
    pub token_type: TokenType,
    /// The text reference from the source code
    pub span: Span,
    /// The text reference from the source code
    pub origin: String,
    /// The line number where the token was parsed from
    pub line: usize,
}

/// `TokenType` are the valid tokens types that lox code can be lexed into.
#[derive(PartialEq, Debug, Clone, Copy)]
#[doc(hidden)]
pub enum TokenType {
    // single-character
    RightParen,
    LeftParen,
    RightBrace,
    LeftBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // one or two character tokens
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // literals
    String,
    Identifier,
    Number(f64),
    // Keywords.
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
    // Misc
    Eof,
}

/// `Lexer` is responsible for iterating over a input string and emitting `Token` for
/// each detected `TokenType`. It maintains the following state:
///
/// `rest` - which is a reference to the substring of `whole` which remained to be lexed.
/// `byte` - the current offset of the byte within the input string being lexed.
/// `line_num` - the current line being processed by the lexer. 1-indexed, used for error reporting.
pub struct Lexer<'de> {
    whole: &'de str,
    rest: &'de str,
    byte: usize,
    line_num: usize,
    at_eof: bool,
}

impl<'de> Lexer<'de> {
    /// Create a new instance of `Lexer` for the given input str.
    #[must_use]
    pub const fn new(input: &'de str) -> Self {
        Self {
            rest: input,
            whole: input,
            byte: 0,
            line_num: 1,
            at_eof: false,
        }
    }
    fn match_reserved_word(c_str: &str) -> TokenType {
        match c_str {
            "and" => TokenType::And,
            "class" => TokenType::Class,
            "else" => TokenType::Else,
            "false" => TokenType::False,
            "fun" => TokenType::Fun,
            "for" => TokenType::For,
            "if" => TokenType::If,
            "nil" => TokenType::Nil,
            "or" => TokenType::Or,
            "print" => TokenType::Print,
            "return" => TokenType::Return,
            "super" => TokenType::Super,
            "this" => TokenType::This,
            "true" => TokenType::True,
            "var" => TokenType::Var,
            "while" => TokenType::While,
            _ => TokenType::Identifier,
        }
    }
}

enum Started {
    String,
    Number,
    Identifier,
    IfNextEqual(TokenType, TokenType),
    CodeComment,
}

impl Iterator for Lexer<'_> {
    type Item = Result<Token, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.at_eof {
            return None;
        }
        loop {
            let mut chars = self.rest.char_indices();
            if let Some(x) = chars.next() {
                let (at, c) = x;
                let mut c_str = &self.rest[at..at + c.len_utf8()];
                let c_onwards = self.rest;
                self.rest = chars.as_str();
                let start_byte = self.byte;
                self.byte += c.len_utf8();

                let started = match c {
                    '(' => {
                        return Some(Ok(Token {
                            token_type: TokenType::LeftParen,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    ')' => {
                        return Some(Ok(Token {
                            token_type: TokenType::RightParen,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '{' => {
                        return Some(Ok(Token {
                            token_type: TokenType::LeftBrace,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '}' => {
                        return Some(Ok(Token {
                            token_type: TokenType::RightBrace,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    ',' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Comma,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '.' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Dot,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '-' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Minus,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '+' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Plus,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '*' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Star,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    ';' => {
                        return Some(Ok(Token {
                            token_type: TokenType::Semicolon,
                            span: Span { start: start_byte, end: start_byte},
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }))
                    }
                    '"' => Started::String,
                    '0'..='9' => Started::Number,
                    'a'..='z' | 'A'..='Z' | '_' => Started::Identifier,
                    '<' => Started::IfNextEqual(TokenType::LessEqual, TokenType::Less),
                    '>' => Started::IfNextEqual(TokenType::GreaterEqual, TokenType::Greater),
                    '!' => Started::IfNextEqual(TokenType::BangEqual, TokenType::Bang),
                    '=' => Started::IfNextEqual(TokenType::EqualEqual, TokenType::Equal),
                    '/' => Started::CodeComment,
                    c if c.is_whitespace() => {
                        if c == '\n' {
                            self.line_num += 1;
                        }
                        continue;
                    }
                    _ => {
                        return Some(Err(miette! {
                            labels = vec![LabeledSpan::at(self.byte-c.len_utf8()..self.byte, "this character")],
                            "[line {}] Error: Unexpected character: {c}", self.line_num
                        }
                        .with_source_code(self.whole.to_string())))
                    }
                };
                match started {
                    Started::CodeComment => {
                        if self.rest.starts_with('/') {
                            let eol = self.rest.find('\n');
                            if let Some(idx) = eol {
                                self.byte += idx;
                                self.rest = &self.rest[idx..];
                            } else {
                                self.byte = self.whole.len();
                                self.rest = &self.rest[self.rest.len()..self.rest.len()];
                            }
                        } else {
                            return Some(Ok(Token {
                                token_type: TokenType::Slash,
                                span: Span {
                                    start: start_byte,
                                    end: self.byte,
                                },
                                origin: c_str.to_string(),
                                line: self.line_num,
                            }));
                        }
                    }
                    Started::String => {
                        if self.rest.contains('"') {
                            let (s, _) = self.rest.split_once('"')?;
                            let c_str = &c_onwards[1..=s.len()];
                            self.byte += s.len() + 1;
                            self.rest = &self.rest[s.len() + 1..];
                            return Some(Ok(Token {
                                token_type: TokenType::String,
                                span: Span {
                                    start: start_byte,
                                    end: self.byte - 1,
                                },
                                origin: c_str.to_string(),
                                line: self.line_num,
                            }));
                        }
                        // Scan to end, we cannot continue to scan for tokens when in an unterminated string
                        self.byte = self.whole.len();
                        self.rest = &self.rest[self.rest.len()..self.rest.len()];
                        return Some(Err(miette! {
                            labels = vec![LabeledSpan::at(start_byte..self.byte, "this unterminated string")],
                            "[line {}] Error: Unterminated string.", self.line_num
                        }
                        .with_source_code(self.whole.to_string())));
                    }
                    Started::Number => loop {
                        let next_num = chars.next();
                        if let Some((_, cn)) = next_num {
                            if cn.is_numeric() {
                                c_str = &c_onwards[at..c_str.len() + cn.len_utf8()];
                                self.rest = chars.as_str();
                                self.byte += cn.len_utf8();
                            } else if cn == '.' {
                                if let Some((_, c_peek)) = chars.next() {
                                    // 456. != 456.0 but unstead 456 DOT
                                    if c_peek.is_numeric() {
                                        c_str = &c_onwards
                                            [at..c_str.len() + cn.len_utf8() + c_peek.len_utf8()];
                                        self.rest = chars.as_str();
                                        self.byte += cn.len_utf8() + c_peek.len_utf8();
                                    } else {
                                        let num = c_str.parse().expect(
                                            "We have called is_numeric on each char in `c_str`",
                                        );
                                        return Some(Ok(Token {
                                            token_type: TokenType::Number(num),
                                            span: Span {
                                                start: start_byte,
                                                end: self.byte - 1,
                                            },
                                            origin: c_str.to_string(),
                                            line: self.line_num,
                                        }));
                                    }
                                }
                            } else {
                                let num: f64 = c_str
                                    .parse()
                                    .expect("We have called is_numeric on each in char in `c_str`");
                                return Some(Ok(Token {
                                    token_type: TokenType::Number(num),
                                    span: Span {
                                        start: start_byte,
                                        end: self.byte - 1,
                                    },
                                    origin: c_str.to_string(),
                                    line: self.line_num,
                                }));
                            }
                        } else {
                            let num: f64 = c_str
                                .parse()
                                .expect("We have called is_numeric on each in char in `c_str`");
                            return Some(Ok(Token {
                                token_type: TokenType::Number(num),
                                span: Span {
                                    start: start_byte,
                                    end: self.byte - 1,
                                },
                                origin: c_str.to_string(),
                                line: self.line_num,
                            }));
                        }
                    },
                    Started::Identifier => loop {
                        let next_char = chars.next();

                        if let Some((_, cn)) = next_char {
                            if cn == ' ' || cn == '\n' {
                                let token_type = Self::match_reserved_word(c_str);
                                return Some(Ok(Token {
                                    token_type,
                                    span: Span {
                                        start: start_byte,
                                        end: self.byte - 1,
                                    },
                                    origin: c_str.to_string(),
                                    line: self.line_num,
                                }));
                            }
                            if cn.is_alphanumeric() || cn == '_' {
                                c_str = &c_onwards[at..c_str.len() + cn.len_utf8()];
                                self.rest = chars.as_str();
                                self.byte += cn.len_utf8();
                                continue;
                            }
                            let token_type = Self::match_reserved_word(c_str);
                            return Some(Ok(Token {
                                token_type,
                                span: Span {
                                    start: start_byte,
                                    end: self.byte - 1,
                                },
                                origin: c_str.to_string(),
                                line: self.line_num,
                            }));
                        }
                        let token_type = Self::match_reserved_word(c_str);
                        return Some(Ok(Token {
                            token_type,
                            span: Span {
                                start: start_byte,
                                end: self.byte - 1,
                            },
                            origin: c_str.to_string(),
                            line: self.line_num,
                        }));
                    },
                    Started::IfNextEqual(then, else_t) => {
                        if self.rest.starts_with('=') {
                            self.byte += 1;
                            self.rest = &self.rest[1..];
                            let token = Token {
                                token_type: then,
                                span: Span {
                                    start: start_byte,
                                    end: self.byte - 1,
                                },
                                origin: c_str.to_string(),
                                line: self.line_num,
                            };
                            return Some(Ok(token));
                        }
                        let token = Token {
                            token_type: else_t,
                            span: Span {
                                start: start_byte,
                                end: self.byte - 1,
                            },
                            origin: c_str.to_string(),
                            line: self.line_num,
                        };
                        return Some(Ok(token));
                    }
                }
            } else {
                self.at_eof = true;
                return Some(Ok(Token {
                    token_type: TokenType::Eof,
                    span: Span {
                        start: self.byte,
                        end: self.byte,
                    },
                    origin: String::new(),
                    line: self.line_num,
                }));
            }
        }
    }
}
