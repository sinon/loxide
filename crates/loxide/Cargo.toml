[package]
name = "loxide"
version.workspace = true
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
keywords.workspace = true
homepage.workspace = true
repository.workspace = true

[dependencies]
clap = { version = "4.5.23", features = ["derive"] }
lexer = { path = "../lexer" }
parser = { path = "../parser" }
ast = { path = "../ast" }
miette = { workspace = true }

[dev-dependencies]
rstest = "0.23.0"
assert_cmd = "2.0.14"
insta = { version = "1.41.1", features = ["yaml"] }
