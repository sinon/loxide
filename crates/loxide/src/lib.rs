//! Loxide
//!
//! The Lexing and Parsing logic to implement a interpreter for the Lox programming language
//!
//! As defined in <https://craftinginterpreters.com>

#![allow(clippy::float_cmp)]
#![allow(clippy::too_many_lines)]
#![warn(missing_docs)]
mod builtins;
pub mod interpreter;
