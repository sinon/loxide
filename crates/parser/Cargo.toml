[package]
name = "parser"
version.workspace = true
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
keywords.workspace = true
homepage.workspace = true
repository.workspace = true

[dependencies]
lexer = { path = "../lexer" }


[dev-dependencies]
rstest = "0.23.0"
insta = { version = "1.41.1", features = ["yaml"] }
