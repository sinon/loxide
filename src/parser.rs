//! Parser
//!
//! Responsible for transforming a given token stream into an AST
//!
//! Uses a recursive desecent parser. To transform the token stream into
//! `Expr`
use crate::lexer::{Lexer, Token, TokenType};

/// `Parser` is responsible for iterating over the token stream from `Lexer`
/// and converting the lexed `Token` into `Expr` which represent an Abstract Syntax Tree (AST)
pub struct Parser<'de> {
    tokens: Vec<Token<'de>>,
    current: usize,
    parse_failed: bool,
}

#[derive(Debug, Clone, PartialEq)]
/// `LiteralAtom` represents the types of literals supported by Lox
pub enum LiteralAtom<'de> {
    /// `String` literal for example `"foo"`
    String(&'de str),
    /// Number literal for example `123.1`
    Number(f64),
    /// Nil literal
    Nil,
    /// Bool literals `false` or `true`
    Bool(bool),
}

#[derive(Debug)]
/// `Expr` represents a unit of an AST
pub enum Expr<'de> {
    /// `Binary` is a binary expression such as `1 * 2`
    Binary {
        /// The left item `Expr` in an expression
        left: Box<Expr<'de>>,
        /// The operator to be applied on the `left` and `right` `Expr`
        operator: Token<'de>,
        /// The right item `Expr` in an expression.
        right: Box<Expr<'de>>,
    },
    /// `Unary` is a unary expression such as `!true`
    Unary {
        /// The operator to be applied on the `right` `Expr`
        operator: Token<'de>,
        /// The expression the unary operator will be applied to
        right: Box<Expr<'de>>,
    },
    /// `Literal` is a value
    Literal(LiteralAtom<'de>),
    /// `Grouping` holds other `Expr` such as `(1 * 2)`
    Grouping(Box<Expr<'de>>),
    /// `Variable`
    Variable(Token<'de>),
    /// `Assign`
    Assign(&'de str, Box<Expr<'de>>),
    /// `Logical` - `or` and `and`
    Logical {
        /// The left expression of a Logical expression
        left: Box<Expr<'de>>,
        /// The operator of a Logical expression
        operator: Token<'de>,
        /// The right expression of a Logical expression
        right: Box<Expr<'de>>,
    },
    /// Function `Call`
    Call {
        /// function to be called
        callee: Box<Expr<'de>>,
        /// paren token
        paren: Token<'de>,
        /// arguments to be passed to function call
        arguments: Vec<Expr<'de>>,
    },
}

#[derive(Debug)]
/// `Stmt` represents the possible statements supported
pub enum Stmt<'de> {
    /// A print statement
    Print(Expr<'de>),
    /// An expression statement
    ExpressionStatement(Expr<'de>),
    /// Var statement
    // var <identifier> = expr;
    Var(&'de str, Option<Expr<'de>>),
    /// Block
    Block(Vec<Stmt<'de>>),
    /// If statement
    If(Expr<'de>, Box<Stmt<'de>>, Option<Box<Stmt<'de>>>),
    /// While statement
    While {
        /// The condition that must be `true` for the body to be run
        condition: Expr<'de>,
        /// The statements that will be executed repreatedly if `condition`
        body: Box<Stmt<'de>>,
    },
}

impl<'de> Iterator for Parser<'de> {
    type Item = Result<Stmt<'de>, String>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.is_at_end() || self.parse_failed {
            return None;
        }

        let stmt = self.declaration();
        match stmt {
            Ok(s) => Some(Ok(s)),
            Err(err) => {
                eprintln!("{err}");
                self.parse_failed = true;
                Some(Err(err))
            }
        }
    }
}

impl<'de> Parser<'de> {
    /// Create new `Parser` from a lexed token stream
    #[must_use]
    pub fn new(input: &'de str) -> Self {
        let tokens: Vec<Token<'_>> = Lexer::new(input).flatten().collect();
        Parser {
            tokens,
            current: 0,
            parse_failed: false,
        }
    }

    fn declaration(&mut self) -> Result<Stmt<'de>, String> {
        if self.match_tokens(&[TokenType::Var]) {
            return self.var_declaration();
        }
        self.statement()
    }

    fn var_declaration(&mut self) -> Result<Stmt<'de>, String> {
        let name = self
            .consume(&TokenType::Identifier, "Expect variable name.")?
            .origin;
        let intializer = if self.match_tokens(&[TokenType::Equal]) {
            Some(self.expression()?)
        } else {
            None
        };
        self.consume(&TokenType::Semicolon, "Expect ';' after value.")?;
        Ok(Stmt::Var(name, intializer))
    }

    fn statement(&mut self) -> Result<Stmt<'de>, String> {
        if self.match_tokens(&[TokenType::While]) {
            return self.while_statement();
        }
        if self.match_tokens(&[TokenType::For]) {
            return self.for_statement();
        }
        if self.match_tokens(&[TokenType::If]) {
            return self.if_statement();
        }
        if self.match_tokens(&[TokenType::Print]) {
            return self.print_statement();
        }
        if self.match_tokens(&[TokenType::LeftBrace]) {
            let stmts = self.block()?;
            return Ok(Stmt::Block(stmts));
        }
        self.expression_statement()
    }

    fn for_statement(&mut self) -> Result<Stmt<'de>, String> {
        self.consume(&TokenType::LeftParen, "Expect '(' after 'for'")?;

        let mut initializer: Option<Stmt> = None;
        if self.match_tokens(&[TokenType::Semicolon]) {
        } else if self.match_tokens(&[TokenType::Var]) {
            initializer = Some(self.var_declaration()?);
        } else {
            initializer = Some(self.expression_statement()?);
        }

        let condition = if self.check(&TokenType::Semicolon) {
            Expr::Literal(LiteralAtom::Bool(true))
        } else {
            self.expression()?
        };
        self.consume(&TokenType::Semicolon, "Expect ';' after loop condition.")?;

        let increment = if self.check(&TokenType::RightParen) {
            None
        } else {
            Some(self.expression()?)
        };
        self.consume(&TokenType::RightParen, "Expect ')' after for clauses.")?;

        let mut body = self.statement()?;

        if let Some(inc) = increment {
            body = Stmt::Block(vec![body, Stmt::ExpressionStatement(inc)]);
        }
        body = Stmt::While {
            condition,
            body: Box::new(body),
        };
        if let Some(init) = initializer {
            body = Stmt::Block(vec![init, body]);
        }
        Ok(body)
    }

    fn while_statement(&mut self) -> Result<Stmt<'de>, String> {
        self.consume(&TokenType::LeftParen, "Expect '(' after 'while'")?;
        let condition = self.expression()?;
        self.consume(&TokenType::RightParen, "Expect ')' after condition")?;
        let body = Box::new(self.statement()?);

        Ok(Stmt::While { condition, body })
    }

    fn if_statement(&mut self) -> Result<Stmt<'de>, String> {
        self.consume(&TokenType::LeftParen, "Expect ')' after 'if'")?;
        let condition = self.expression()?;
        self.consume(&TokenType::RightParen, "Expect ')' after if condition.")?;

        let then_branch = self.statement()?;
        let else_branch = if self.match_tokens(&[TokenType::Else]) {
            Some(Box::new(self.statement()?))
        } else {
            None
        };
        let if_stmt = Stmt::If(condition, Box::new(then_branch), else_branch);
        Ok(if_stmt)
    }

    fn block(&mut self) -> Result<Vec<Stmt<'de>>, String> {
        let mut stmts: Vec<Stmt<'de>> = Vec::new();
        while !self.check(&TokenType::RightBrace) && !self.is_at_end() {
            stmts.push(self.declaration()?);
        }

        self.consume(&TokenType::RightBrace, "Expect '}' after block.")?;
        Ok(stmts)
    }

    fn expression_statement(&mut self) -> Result<Stmt<'de>, String> {
        let expr = self.expression()?;
        self.consume(&TokenType::Semicolon, "Expect ';' after value.")?;
        Ok(Stmt::ExpressionStatement(expr))
    }

    fn print_statement(&mut self) -> Result<Stmt<'de>, String> {
        let expr = self.expression()?;
        self.consume(&TokenType::Semicolon, "Expect ';' after value.")?;
        Ok(Stmt::Print(expr))
    }

    fn expression(&mut self) -> Result<Expr<'de>, String> {
        self.assignment()
    }

    fn assignment(&mut self) -> Result<Expr<'de>, String> {
        let expr = self.or()?;

        if self.match_tokens(&[TokenType::Equal]) {
            let equals = self.previous().clone();
            let value = self.assignment()?;

            if let Expr::Variable(token) = expr {
                let name = token.origin;
                return Ok(Expr::Assign(name, Box::new(value)));
            }
            let err_msg = Self::error_msg(
                &equals.token_type,
                equals.origin,
                equals.line,
                "Invalid assignment type.",
            );
            return Err(err_msg);
        }

        Ok(expr)
    }

    fn or(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.and()?;

        while self.match_tokens(&[TokenType::Or]) {
            let operator = self.previous().clone();
            let right = self.and()?;
            expr = Expr::Logical {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expr)
    }

    fn and(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.equality()?;
        while self.match_tokens(&[TokenType::And]) {
            let operator = self.previous().clone();
            let right = self.equality()?;
            expr = Expr::Logical {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            }
        }
        Ok(expr)
    }

    fn equality(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.comparison()?;

        while self.match_tokens(&[TokenType::BangEqual, TokenType::EqualEqual]) {
            let operator = self.previous().clone();
            let right = self.comparison()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expr)
    }

    fn comparison(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.term()?;

        while self.match_tokens(&[
            TokenType::Greater,
            TokenType::GreaterEqual,
            TokenType::Less,
            TokenType::LessEqual,
        ]) {
            let operator = self.previous().clone();
            let right = self.term()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expr)
    }

    fn term(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.factor()?;

        while self.match_tokens(&[TokenType::Minus, TokenType::Plus]) {
            let operator = self.previous().clone();
            let right = self.factor()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expr)
    }

    fn factor(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.unary()?;

        while self.match_tokens(&[TokenType::Slash, TokenType::Star]) {
            let operator = self.previous().clone();
            let right = self.unary()?;
            expr = Expr::Binary {
                left: Box::new(expr),
                operator,
                right: Box::new(right),
            };
        }

        Ok(expr)
    }

    fn unary(&mut self) -> Result<Expr<'de>, String> {
        if self.match_tokens(&[TokenType::Bang, TokenType::Minus]) {
            let operator = self.previous().clone();
            let right = self.unary()?;
            return Ok(Expr::Unary {
                operator,
                right: Box::new(right),
            });
        }

        self.call()
    }

    fn call(&mut self) -> Result<Expr<'de>, String> {
        let mut expr = self.primary()?;

        loop {
            if self.match_tokens(&[TokenType::LeftParen]) {
                expr = self.finish_call(expr)?;
            } else {
                break;
            }
        }
        Ok(expr)
    }

    fn finish_call(&mut self, callee: Expr<'de>) -> Result<Expr<'de>, String> {
        let mut arguments: Vec<Expr<'de>> = vec![];

        if !self.check(&TokenType::RightParen) {
            loop {
                arguments.push(self.expression()?);
                // https://users.rust-lang.org/t/how-many-arguments-can-i-pass-to-a-function/84250
                if arguments.len() >= 65535 {
                    return Err("Can't have more than 65535 arguments.".to_string());
                }
                if !self.match_tokens(&[TokenType::Comma]) {
                    break;
                }
            }
        }
        let paren = self
            .consume(&TokenType::RightParen, "Expect ')' after arguments.")?
            .clone();

        Ok(Expr::Call {
            callee: Box::new(callee),
            paren,
            arguments,
        })
    }

    fn primary(&mut self) -> Result<Expr<'de>, String> {
        let token = self.advance();
        match token.token_type {
            TokenType::Number(n) => Ok(Expr::Literal(LiteralAtom::Number(n))),
            TokenType::String => Ok(Expr::Literal(LiteralAtom::String(token.origin))),
            TokenType::True => Ok(Expr::Literal(LiteralAtom::Bool(true))),
            TokenType::False => Ok(Expr::Literal(LiteralAtom::Bool(false))),
            TokenType::Nil => Ok(Expr::Literal(LiteralAtom::Nil)),
            TokenType::LeftParen => {
                let expr = self.expression()?;
                self.consume(&TokenType::RightParen, "Expect ')' after expression")?;
                Ok(Expr::Grouping(Box::new(expr)))
            }
            TokenType::Identifier => Ok(Expr::Variable(self.previous().clone())),
            _ => {
                let err_msg = Self::error_msg(
                    &token.token_type,
                    token.origin,
                    token.line,
                    "Expect expression",
                );
                Err(err_msg)
            }
        }
    }

    fn match_tokens(&mut self, types: &[TokenType]) -> bool {
        for t in types {
            if self.check(t) {
                self.advance();
                return true;
            }
        }
        false
    }

    fn check(&self, token_type: &TokenType) -> bool {
        if self.is_at_end() {
            return false;
        }
        std::mem::discriminant(&self.peek().token_type) == std::mem::discriminant(token_type)
    }

    fn advance(&mut self) -> &Token<'de> {
        if !self.is_at_end() {
            self.current += 1;
        }
        self.previous()
    }

    fn peek(&self) -> &Token<'de> {
        self.tokens
            .get(self.current)
            .expect("current should never be outside index")
    }

    fn previous(&self) -> &Token<'de> {
        self.tokens.get(self.current - 1).unwrap()
    }

    fn is_at_end(&self) -> bool {
        let token = self.peek();
        matches!(token.token_type, TokenType::Eof)
    }

    fn consume(&mut self, token_type: &TokenType, message: &str) -> Result<&Token<'de>, String> {
        if self.check(token_type) {
            Ok(self.advance())
        } else {
            let token = self.peek();
            Err(Self::error_msg(
                token_type,
                token.origin,
                token.line,
                message,
            ))
        }
    }

    fn error_msg(token_type: &TokenType, origin: &str, num: usize, message: &str) -> String {
        if *token_type == TokenType::Eof {
            format!("[line {num}] Error at end: {message}.")
        } else {
            format!("[line {num}] Error at '{origin}': {message}.")
        }
    }
}
