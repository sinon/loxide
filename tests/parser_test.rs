use insta::assert_snapshot;
use loxide::lexer::{Lexer, Token};
use loxide::parser::Parser;
use rstest::*;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        settings.set_description(format!($($expr,)*));
        settings.set_omit_expression(true);
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
#[case("true")]
#[case("false")]
#[case("nil")]
fn test_parser_literals(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    assert_snapshot!(exprs.join("\n"));
}

#[rstest]
#[case("(\"foo\")")]
#[case("(1)")]
#[case("((true))")]
fn test_parser_parentheses(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    assert_snapshot!(exprs.join("\n"));
}

#[rstest]
#[case("!true")]
#[case("!false")]
#[case("-10")]
fn test_parser_unary(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    insta::with_settings!({
        description => input, // the template source code
        omit_expression => true // do not include the default expression
    }, {
        assert_snapshot!(exprs.join("\n"));
    });
}

#[rstest]
#[case("16 * 38 / 58")]
#[case("52 + 80 - 94")]
fn test_parser_arimethic(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    assert_snapshot!(exprs.join("\n"));
}

#[rstest]
#[case("83 < 99 > 115")]
#[case("52 <= 80 >= 94")]
fn test_parser_comparison(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();

    insta::with_settings!({
        description => input, // the template source code
        omit_expression => true // do not include the default expression
    }, {
        assert_snapshot!(exprs.join("\n"));
    });
}

#[rstest]
#[case("\"baz\" == \"baz\"")]
#[case("10 != 9")]
fn test_parser_equality(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(input).filter_map(Result::ok).collect();
    let exprs: Vec<String> = Parser::new(tokens)
        .take_while(|x| x.is_ok())
        .map(|e| match e {
            Ok(exp) => format!("{}", exp),
            Err(_) => "".to_string(),
        })
        .collect();
    assert_snapshot!(exprs.join("\n"));
}

#[rstest]
#[case("(72 +)")]
#[case("\"foo")]
fn test_parser_error(#[case] input: &str) {
    set_snapshot_suffix!("{}", input);
    let tokens: Vec<Token> = Lexer::new(&input).filter_map(Result::ok).collect();
    for expr in Parser::new(tokens) {
        match expr {
            Ok(_) => todo!(),
            Err(err) => {
                assert_snapshot!(err);
                break;
            }
        }
    }
    // let exprs: Vec<String> = Parser::new(tokens)
    //     .take_while(|x| !x.is_err())
    //     .map(|e| match e {
    //         Ok(exp) => format!("{}", exp),
    //         Err(err) => panic!("!"),
    //     })
    //     .collect();
    // assert_eq!(
    //     exprs.join("\n"),
    //     "[line 1] Error at ')': Expect expression."
    // )
}
