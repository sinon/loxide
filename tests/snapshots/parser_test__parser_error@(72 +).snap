---
source: tests/parser_test.rs
description: (72 +)
---
[line 1] Error at ')': Expect expression.
